text/0000-fabric-private-chaincode-1.0.md

```diff
@@ -0,0 +1,720 @@
+Notably, none of these features (or relative cryptographic keys) are exposed to the FPC Chaincode.
+See [Appendix](#appendix) for more details.
+During an FPC transaction invocation, the FPC Shim inside the enclave represents the endpoint of the secure channel between the FPC Client and the FPC Chaincode.
+Hence, at the lower level of the protocol stack, the Fabric client and the peer only handle encrypted and integrity protected transactional information.
+Inside the Chaincode Enclave, a C++ based FPC Shim is available to FPC Chaincode developers.
+This shim follows the programming model of the Fabric Go shim, though using a different language.
+Initially, the FPC Shim comprises a subset of the standard Fabric Shim and is complemented in the future.
+These details are documented separately in the Shim header file itself: **[ecc_enclave/enclave/shim.h](https://github.com/hyperledger-labs/fabric-private-chaincode/blob/master/ecc_enclave/enclave/shim.h)**
+
+### Enclave Registry
+
+The Enclave Registry helps to establish trust in the enclave and the secure channel.
+It is implemented as a chaincode which maintains a list of all Chaincode Enclaves deployed on the peers in a channel.
+The registry associates with each enclave their identity, associated public keys and an attestation linking them.
+In detail, after an FPC Chaincode definition is committed on the channel, the chaincode's hosting enclave must be registered by invoking the Enclave Registry, in order to become operational.
+When registering a new enclave, the enclave registry verifies the provided enclave attestation (signed by the trusted hardware manufacturer) and checks it against the chaincode definition. In particular, the enclave attestation contains the code identity (`MRENCLAVE`) that must match the `version` field of the chaincode definition (recall deployment section earlier).
+If the verification succeeds, the enclave registry stores the attested data (including the public keys) on the ledger and thereby make it available for members of the channel.
+<!--  -->
+This enables any channel member to inspect an enclave attestation to verify the genuinity of the enclave in order to establish trust in its public keys.
+
+Additionally, the registry manages chaincode specific keys, including a chaincode public encryption key, and facilitates corresponding key-management among authorized chaincode enclaves.
+The registry is particularly relevant for clients, for retrieving the encryption key to protect the transaction arguments and set up a direct secure channel.
+Note that clients can access the enclave registry using "strong queries" where multiple peers are queried in order to get a correct response and not relying on a single peer.
+This prevents a malicious peer from responding with a wrong transaction encryption key.
+
+
+
+### Enclave Endorsement Validation
```

Comment:
The paragraph makes it sound like this runs at peer validation time, e.g. "applies any state updates". But I believe it actually runs at endorsement time. Please clarify.

This becomes more clear below, but I think you should conceptually summarize the two-step execution process here as an introduction (first the FPC chaincode executes, and then the validation chaincode executes, with the latter serving as the traditional Fabric transaction).

Reply:
I rephrased for clarification and added a forward reference.